From aacfafe8713427f48a7d948ad57e6967692e519a Mon Sep 17 00:00:00 2001
From: nikhilad <nikhilad@ami.com>
Date: Wed, 29 Oct 2025 16:28:10 +0530
Subject: [PATCH] Add empty EEPROM Fru Update Support

This commit add support to identify empty EEPROM and assign
FRU Device ID.

Tested in Intel Archercity:

1. Confirm the EEPROM is empty using i2ctool commands

2. List the FRU and confirm Empty FRU is listing
root@AMIOT-4ecb6735a597:~# ipmitool fru list
FRU Device Description : Builtin FRU Device (ID 0)
 Chassis Type          : Rack Mount Chassis
 Chassis Part Number   : ..................
 Chassis Serial        : ..................
 Chassis Extra         : ...............................
 Chassis Extra         : ...............................
 Board Mfg Date        : Mon Jan  1 00:00:00 1996
 Board Mfg             : Intel Corporation
 Board Product         : ArcherCity
 Board Serial          : Intel_3245
 Board Part Number     : ..........
 Product Manufacturer  : Intel Corporation
 Product Name          : ArcherCity
 Product Part Number   : ..........
 Product Version       : ....................
 Product Serial        : Intel_3245
 Product Asset Tag     : ....................

FRU Device Description : UNKNOWN (ID 102)
 Invalid FRU size 0

3. Now Try to update the FRU using IPMI command

ipmitool raw 0x0a 0x12 0x66 0x00 0x00 0x01 0x00 0x00 0x00 0x01
0x0d 0x00 0xf1 0x01 0x0c 0x19 0xe1 0x53 0x4f 0x4c 0x55 0x4d 0x20
0x43 0x4f 0x2e 0x2c 0x20 0x4c 0x54 0x44 0x2e 0x20 0x20 0x20 0x20
0x20 0x20 0x20 0x20 0x20 0x20 0x20 0x20 0x20 0x20 0x20 0x20 0x20
0x20 0xc8 0x49 0x53 0x31 0x36 0x32 0x46 0x32 0x32 0xca 0x47 0x33
0x36 0x32 0x33 0x34 0x2d 0x30 0x31 0x36 0xc3 0x31 0x31 0x41 0xdc
0x50 0x53 0x53 0x46 0x31 0x36 0x32 0x32 0x30 0x32 0x41 0x43 0x4e
0x53 0x31 0x36 0x32 0x32 0x41 0x34 0x42 0x4c 0x33 0x50 0x31 0x31
0x33 0x32 0x00 0x00 0xc1 0x00 0x00 0xc4 0x00 0x02 0x18 0x43 0xa3
0x40 0x06 0x34 0x08 0x41 0x02 0x28 0x23 0xb0 0x36 0x50 0x46 0x20
0x67 0x2f 0x3f 0x0c 0x0e 0x34 0xa8 0x00 0x40 0x06 0x00 0x01 0x02
0x0d 0xe6 0x0a 0x01 0xb0 0x04 0x74 0x04 0xec 0x04 0x78 0x00 0x00
0x00 0x85 0x00 0x01 0x82 0x0d 0x31 0x3f 0x82 0xb0 0x04 0x74 0x04
0xec 0x04 0x78 0x00 0x00 0x00 0xac 0x0d

4. Now List the FRU and confirm its updated

root@AMIOT-4ecb6735a597:~# ipmitool fru list
FRU Device Description : Builtin FRU Device (ID 0)
 Chassis Type          : Rack Mount Chassis
 Chassis Part Number   : ..................
 Chassis Serial        : ..................
 Chassis Extra         : ...............................
 Chassis Extra         : ...............................
 Board Mfg Date        : Mon Jan  1 00:00:00 1996
 Board Mfg             : Intel Corporation
 Board Product         : ArcherCity
 Board Serial          : Intel_3245
 Board Part Number     : ..........
 Product Manufacturer  : Intel Corporation
 Product Name          : ArcherCity
 Product Part Number   : ..........
 Product Version       : ....................
 Product Serial        : Intel_3245
 Product Asset Tag     : ....................

FRU Device Description : IS162F22 (ID 84)
 Product Manufacturer  : SOLUM CO., LTD.
 Product Name          : IS162F22
 Product Part Number   : G36234-016
 Product Version       : 11A
 Product Serial        : PSSF162202ACNS1622A4BL3P1132

Signed-off-by: vipinc <vipinc@ami.com>

Signed-off-by: nikhilad <nikhilad@ami.com>
---
 src/fru_device/fru_device.cpp | 189 ++++++++++++++++++++++++++++++++--
 1 file changed, 183 insertions(+), 6 deletions(-)

diff --git a/src/fru_device/fru_device.cpp b/src/fru_device/fru_device.cpp
index 64b1e9a..960ca68 100644
--- a/src/fru_device/fru_device.cpp
+++ b/src/fru_device/fru_device.cpp
@@ -254,6 +254,46 @@ static void makeProbeInterface(size_t bus, size_t address,
     it->second->initialize();
 }
 
+static bool isFruConfigured(uint8_t bus, uint8_t adress)
+{
+    std::string configFile =
+        "/usr/share/entity-manager/configurations/eeprom.json";
+    using Json = nlohmann::json;
+    std::ifstream jsonFile(configFile);
+    if (!jsonFile.good())
+    {
+        std::cerr << "JSON file not found " << configFile << "\n";
+        return false;
+    }
+
+    Json data = nullptr;
+    try
+    {
+        data = Json::parse(jsonFile, nullptr, false);
+    }
+    catch (const Json::parse_error& e)
+    {
+        std::cerr << "Faile to parse FRU Json file";
+        return false;
+    }
+
+    if (data.empty())
+    {
+        return false;
+    }
+    for (const auto& fruConfigs : data["FRU_EEPROM"])
+    {
+        uint8_t busConf = static_cast<uint8_t>(fruConfigs["Bus"]);
+        uint8_t addrConf = static_cast<uint8_t>(fruConfigs["Address"]);
+        if ((busConf == bus) && (addrConf == adress))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+
 // Issue an I2C transaction to first write to_target_buf_len bytes,then read
 // from_target_buf_len bytes.
 static int i2cSmbusWriteThenRead(
@@ -496,7 +536,8 @@ std::set<size_t> findI2CEeproms(int i2cBus,
         foundList.insert(address);
 
         std::vector<uint8_t> device = processEeprom(i2cBus, address);
-        if (!device.empty())
+        if (!device.empty() ||
+            (device.empty() && isFruConfigured(i2cBus, address)))
         {
             devices->emplace(address, device);
         }
@@ -648,7 +689,7 @@ int getBusFRUs(int file, int first, int last, int bus,
                 pair = readFRUContents(readerBytewise, errorMessage);
             }
 
-            if (pair.first.empty())
+            if ((pair.first.empty()) && (!isFruConfigured(bus, ii)))
             {
                 continue;
             }
@@ -863,17 +904,19 @@ void addFruObjectToDbus(
     std::shared_ptr<sdbusplus::asio::connection>& systemBus)
 {
     boost::container::flat_map<std::string, std::string> formattedFRU;
+    std::string productName = "/xyz/openbmc_project/FruDevice/";
 
     std::optional<std::string> optionalProductName = getProductName(
         device, formattedFRU, bus, address, unknownBusObjectCount);
     if (!optionalProductName)
     {
         std::cerr << "getProductName failed. product name is empty.\n";
-        return;
+        productName += "Unknown";
+    }
+    else
+    {
+        productName += optionalProductName.value();
     }
-
-    std::string productName =
-        "/xyz/openbmc_project/FruDevice/" + optionalProductName.value();
 
     std::optional<int> index = findIndexForFRU(dbusInterfaceMap, productName);
     if (index.has_value())
@@ -1098,6 +1141,134 @@ bool writeFRU(uint8_t bus, uint8_t address, const std::vector<uint8_t>& fru)
     return true;
 }
 
+std::vector<uint8_t> getFruConfig(const uint16_t& fruBus,
+                                  const uint8_t& fruAddress)
+{
+    auto bus = std::make_shared<sdbusplus::asio::connection>(io);
+    std::vector<uint8_t> result;
+    using GetSubTreePathsType = std::vector<std::string>;
+
+    auto message = bus->new_method_call(
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetSubTreePaths");
+    message.append("/", 0,
+                   std::array<const char*, 1>{
+                       "xyz.openbmc_project.Inventory.Item.FruConfig"});
+    GetSubTreePathsType idPaths;
+    auto reply = bus->call(message);
+    reply.read(idPaths);
+
+    for (const auto& path : idPaths)
+    {
+        std::map<std::string, DBusValueVariant> properties;
+
+        sdbusplus::message_t getProperties =
+            bus->new_method_call("xyz.openbmc_project.FruDevice", path.c_str(),
+                                 "org.freedesktop.DBus.Properties", "GetAll");
+        getProperties.append("xyz.openbmc_project.Inventory.Item.FruConfig");
+        try
+        {
+            sdbusplus::message_t response = bus->call(getProperties);
+            response.read(properties);
+        }
+        catch (const std::exception& e)
+        {
+            std::cerr << "GetAll Failed";
+            continue;
+        }
+        auto configBus = properties.find("Bus");
+        auto configAddr = properties.find("Address");
+        uint8_t confBus = 0;
+        uint8_t confAddr = 0;
+
+        if (configBus == properties.end() || configAddr == properties.end())
+        {
+            std::cerr << "Failed to find Bus and Address\n";
+            continue;
+        }
+        else
+        {
+            if (auto busValue = std::get_if<uint8_t>(&configBus->second);
+                busValue != nullptr)
+            {
+                if (auto addrValue = std::get_if<uint8_t>(&configAddr->second);
+                    addrValue != nullptr)
+                {
+                    confBus = *busValue;
+                    confAddr = *addrValue;
+                }
+            }
+        }
+
+        if ((static_cast<uint8_t>(fruBus) == confBus) &&
+            (static_cast<uint8_t>(fruAddress) == confAddr))
+        {
+            auto fruIdIter = properties.find("FruId");
+            auto fruSizeIter = properties.find("FruSize");
+            if ((fruIdIter != properties.end()) &&
+                (fruSizeIter != properties.end()))
+            {
+                auto fruIdValue = std::get_if<uint8_t>(&fruIdIter->second);
+                result.push_back(static_cast<uint8_t>(*fruIdValue));
+
+                auto fruSizeValue = std::get_if<uint8_t>(&fruSizeIter->second);
+                result.push_back(static_cast<uint8_t>(*fruSizeValue));
+
+                break;
+            }
+            return std::vector<uint8_t>();
+        }
+    }
+    return result;
+}
+
+void DoFruConfig(sdbusplus::asio::object_server& objServer)
+{
+    std::string configFile =
+        "/usr/share/entity-manager/configurations/eeprom.json";
+    using Json = nlohmann::json;
+    std::ifstream jsonFile(configFile);
+    if (!jsonFile.good())
+    {
+        std::cerr << "JSON file not found " << configFile << "\n";
+        return;
+    }
+
+    Json data = nullptr;
+    try
+    {
+        data = Json::parse(jsonFile, nullptr, false);
+    }
+    catch (const Json::parse_error& e)
+    {
+        std::cerr << "Faile to parse FRU Json file";
+    }
+
+    if (data.empty())
+    {
+        std::cerr << "FRU Config data is empty";
+    }
+    for (const auto& fruConfigs : data["FRU_EEPROM"])
+    {
+        uint8_t busConf = static_cast<uint8_t>(fruConfigs["Bus"]);
+        uint8_t addrConf = static_cast<uint8_t>(fruConfigs["Address"]);
+        std::string fruName = static_cast<std::string>(fruConfigs["Name"]);
+        uint8_t fruId = static_cast<uint8_t>(fruConfigs["FruId"]);
+        uint8_t fruSize = static_cast<uint8_t>(fruConfigs["FruSize"]);
+        std::shared_ptr<sdbusplus::asio::dbus_interface> iface =
+            objServer.add_interface(
+                "/xyz/openbmc_project/FruDevice/" + fruName,
+                "xyz.openbmc_project.Inventory.Item.FruConfig");
+
+        iface->register_property("Bus", busConf);
+        iface->register_property("Address", addrConf);
+        iface->register_property("FruId", fruId);
+        iface->register_property("FruSize", fruSize);
+        iface->initialize();
+    }
+}
+
 void rescanOneBus(
     BusMap& busmap, uint16_t busNum,
     boost::container::flat_map<
@@ -1484,6 +1655,11 @@ int main()
             rescanBusses(busMap, dbusInterfaceMap, unknownBusObjectCount,
                          powerIsOn, objServer, systemBus);
         });
+    /*TODO: Need to enable support once, D-Bus destruction is supported
+        iface->register_method("DoFruConfig", [&]() {
+            DoFruConfig(objServer);
+        });
+    */
     iface->initialize();
 
     std::function<void(sdbusplus::message_t & message)> eventHandler =
@@ -1584,6 +1760,7 @@ int main()
     rescanBusses(busMap, dbusInterfaceMap, unknownBusObjectCount, powerIsOn,
                  objServer, systemBus);
 
+    DoFruConfig(objServer);
     io.run();
     return 0;
 }
-- 
2.34.1

